csharp/src/Microsoft.ML.OnnxRuntime/OrtValue.shared.cs

```diff
@@ -1464,11 +1473,155 @@ protected virtual void Dispose(bool disposing)
             }
 
             Debug.Assert(_handle != IntPtr.Zero);
-            NativeMethods.OrtReleaseValue(_handle);
+            if (_owned)
+            {
+                NativeMethods.OrtReleaseValue(_handle);
+            }
             _handle = IntPtr.Zero;
             _disposed = true;
         }
 
         #endregion
     }
+
+    /// <summary>
+    /// Interface for all readonly methods implemented by OrtValue.
+    /// </summary>
+    public interface IReadOnlyOrtValue
```

Comment:
Note:
The C# bindings only implements an `OrtValue` owned by the application. However, we now need to be able to represent an `OrtValue` that is owned by ORT (so that we can pass it to the user's callback function when compiling). 

So, I created a "readonly" interface that contains all OrtValue getters. This makes it possible to pass a `IReadOnlyOrtValue` ortvalue to the user's callback.

Reply:
ORT passes a const OrtValue* to the user's function. Don't want the user's function to modify it or release it (though release is less of an issue since OrtValue is a shared_ptr)